leetcode/medium/227._Basic_Calculator_II.py

```diff
@@ -0,0 +1,51 @@
+# Link to problem: https://leetcode.com/problems/basic-calculator-ii/description/
+# Loom explaining solution you wrote (5m): 
+# Solution:
+# Description
+# Key ideas
+# To step process, process */ first, +- last.
+# - can be treated as + (-val), +- can be mapped as sum
+# When processing */, use a value stack
+# Difficulty: Medium
+# Code
+
+
+def token_of(s):
+    cur = ""
+    for c in s:
+        if c == " ":
+            if cur:
+                yield int(cur)
+                cur = ""
+        elif c in "+-*/":
+            if cur:
+                yield int(cur)
+            cur = ""
+            yield c
+        else:
+            assert c.isnumeric()
+            cur += c
+    if cur:
+        yield int(cur)
+
+class Solution:
+    def calculate(self, s: str) -> int:
+        itr = token_of(s)
+        vals = [next(itr)]
+        try:
```

Comment:
You should need a try/except and while True loop. 

The pseudocode should be to process each token incrementally and use that to create the result. There will be several if/else/elif statements. 

Similarly `match/case` is not commonly used in Python. Use if/elif statements instead.

Reply:
What do you mean by:

> You should need a try/except and while True loop.
> 
> The pseudocode should be to process each token incrementally and use that to create the result. There will be several if/else/elif statements.

What change are you suggesting here? Are you saying I shouldn't need this?

If I rethink this this function should be:
```python
    def calculate(self, s: str) -> int:
        tokens = tokenize(s) # return []
        vals = [tokens[0]]
        for token, rhs in batched(token[1:], 2):
            token = next(itr)
            rhs = next(itr)
            match token:
                case "+":
                    vals.append(rhs)
                case "-":
                    vals.append(-rhs)
                case "*":
                    vals.append(vals.pop() * rhs)
                case "/":
                    vals.append(int(vals.pop() / rhs))
        return sum(vals)
```

Isn't match case a lot more readable? Hum I am surprised this is discouraged.

Reply:
The change I'm suggesting is something like this 
 
```
     tokens = tokenize(s) # return []
        vals = [tokens.pop()]
        for operator, rhs in batched(token, 2):
            if operator == "+":
                    vals.append(rhs)
            elif operator == "-":
                    vals.append(-rhs)
            elif operator == "*"
                    vals.append(vals.pop() * rhs)
            elif operator == "/":
                    vals.append(int(vals.pop() / rhs))
        return sum(vals)
```

This is more Pythonic and consistent with Python standards.